pkg/costattribution/manager.go

```diff
@@ -138,20 +172,79 @@ func (m *Manager) ActiveSeriesTracker(userID string) *ActiveSeriesTracker {
 }
 
 func (m *Manager) Collect(out chan<- prometheus.Metric) {
+	// Clone both maps to avoid holding the locks while collecting metrics.
 	m.stmtx.RLock()
-	for _, tracker := range m.sampleTrackersByUserID {
-		tracker.Collect(out)
+	sampleTrackersByUserID := maps.Clone(m.sampleTrackersByUserID)
+	m.stmtx.RUnlock()
+
+	m.atmtx.RLock()
+	activeTrackersByUserID := maps.Clone(m.activeTrackersByUserID)
+	m.atmtx.RUnlock()
+
+	for _, tracker := range sampleTrackersByUserID {
+		cardinality, overflown := tracker.cardinality()
+
+		out <- prometheus.MustNewConstMetric(
+			m.sampleTrackerCardinalityDesc,
+			prometheus.GaugeValue,
+			float64(cardinality),
+			tracker.userID,
+		)
+		if overflown {
+			out <- prometheus.MustNewConstMetric(
+				m.sampleTrackerOverflowDesc,
+				prometheus.GaugeValue,
+				1,
```

Comment:
Would be easier to understand this metric if it was 1 or 0 instead of 1 or missing?

Reply:
It is, but we'd also need to pay for that extra cardinality (one metric per user per ingester) that is never needed (we don't expect tenants to be in oveflow), so I thought we'd rather just pay the price at query time doing something like `max by (user) (overflown or 0*caradinality)`